src/compiler/scala/tools/nsc/typechecker/Contexts.scala

```diff
@@ -1555,6 +1558,13 @@ trait Contexts { self: Analyzer =>
     }
   }
 
+  private[Contexts] object isAccessibleLambda extends (Symbol => Boolean) {
+    private[this] var context: Context = null
+    private[this] var prefix: Type = null
+    def setup(c: Context, p: Type): Unit = {context = c ; prefix = p }
+    def apply(sym: Symbol): Boolean = context.isAccessible(sym, prefix, superAccess = false)
+  }
```

Comment:
This would need to be a `ReusableInstance` to be safe against reentrant execution (`isAccessible` could call `info` which could trigger lazy typechecking).

Here's another hot call site that avoids the lambda creation for `Symbol#filter` in an different manner:

https://github.com/scala/scala/blob/1d5e5648b90b65395e16548fd4c09f7e01be9bc5/src/reflect/scala/reflect/internal/Symbols.scala#L2360-L2363

I'd prefer a consistent approach. The range of options:

  - if/else special casing `!isOverloaded` and `NoSymbol`. This entails a little code duplication.
  - use a `ReusableInstance` as a one-element cache for the lambda. This obscures the code somewhat as the lambda logic is removed from where it is used. It also leaks a `Context` / `CompllationUnit` in t long-lived `Global` unless it is also registered as a `perRunCache` to cleanup at the end of the `Run`.
  - Make `Symbol#filter` `@inline final`?

Reply:
The special case for `NoSymbol` appears the be the main thing we need. If we push out out into the calling code to avoid calling this at all, we also can avoid creating `Tuole2(null, NoSymbol)` repeatedly.

Alternative proposal with this: https://github.com/scala/scala/pull/8728